### PR TITLE
Complex type support for socket provider

### PR DESCRIPTION
**Pull request by hoopoepg**
hi
here is implementation of complex types of atomics for socket provider.
used GCC/ICC built-in complex types

updated atomic implementation and sock_ep_atomic_valid functions

thank you

### CODE REVIEW

**Comment by jithinjosepkl** [comment]
Thanks for the PR. It looks good to me. Can you please add the --signoff to the commit?
Also, it would be better if you could squash the commits for a better commit history.

**Comment by goodell** [comment]
Just to be clear, by adding your `Signed-off-by:` you are indicating agreement to the points mentioned in https://github.com/ofiwg/libfabric/blob/master/CONTRIBUTORS

File: prov/sockets/src/sock_atomic.c [hunk 1]
```diff
@@ -602,3 +608,6 @@ struct fi_ops_atomic sock_ep_atomic = {
 	.readwritevalid = sock_ep_atomic_valid,
 	.compwritevalid = sock_ep_atomic_valid,
 };
+
+/* ex: set tabstop=8 noexpandtab: */
+
```
Comment (shefty) [inline prov/sockets/src/sock_atomic.c]: this should be dropped

File: prov/sockets/src/sock_progress.c [hunk 2]
```diff
@@ -941,6 +942,58 @@ static int sock_pe_process_rx_write(struct sock_pe *pe, struct sock_rx_ctx *rx_c
...
+		else							\
+			*_cmp = *_dst;					\
+		break;							\
+									\
+	case FI_CSWAP_NE:						\
+		if (*_cmp != *_dst)					\
+			*_dst = *_src;					\
+		else							\
+			*_cmp = *_dst;					\
+		break;							\
```
Comment (shefty) [inline prov/sockets/src/sock_progress.c]: The atomic-fetch operations are defined to return the initial value to the target.  Are these doing this, or are these changes not for the atomic-fetch ops?  I'm just a little confused trying to figure out what the if-else statements in the cswap options are doing.

**Comment by shefty** [comment]
Please see PR #1244 to ensure that this supports the updated man pages, which were adjusted to align with the sockets provider implementation.

File: prov/sockets/src/sock_atomic.c (hunk 1 above)
Comment (hoopoepg) [inline prov/sockets/src/sock_atomic.c]: will do

File: prov/sockets/src/sock_progress.c (hunk 2 above)
Comment (hoopoepg) [inline prov/sockets/src/sock_progress.c]: 'fetch' part of operation is implemented on level up from current scope, 'fetch & compare & swap' operation should have 2 buffers to receive result: first one for original data, second - for 'swapped' data (as i can see this is only way to check if data was swapped), as i can see from sources both 'result' and 'compare' buffers are sent from target rank to originator

**Comment by hoopoepg** [comment]
i created another pull request: https://github.com/ofiwg/libfabric/pull/1260
to implement new data types support i just copy/pasted code for processing int types (look at macro SOCK_ATOMIC_UPDATE_INT in https://github.com/ofiwg/libfabric/blob/master/prov/sockets/src/sock_progress.c)

**Pull request closed without merging**